filter: Support relative dates for `--min-date` and `--max-date`

benjaminotter opened the pull request:
### Description of proposed changes    
The relative dates are parsed by `numeric_date` which uses `datetime.date.today()` to translate the relative date to an absolute date.

Relative dates are positive duration values following the ISO 8601 duration syntax
e.g. `--min-date 1Y2W5D` for 1 year, 2 weeks and 5 days ago or `--max-date 1D` for yesterday

This also adds a package dependency `isodate` to parse the duration string.

### Related issue(s)  
Fixes #721 

### Testing

Added pytests.

### Post-release

- [ ] update Bioconda recipe with `isodate` dependency
- [ ] update `min_date`/`max_date` descriptions in ncov configuration guide (e.g. [`filter.min_date`](https://github.com/nextstrain/ncov/blob/master/docs/src/reference/configuration.rst?plain=1#L324))

huddlej commented:
Thank you for catching this issue, @benjaminotter. Reading [the ISO 8601 standard more closely](https://www.loc.gov/standards/datetime/iso-tc154-wg5_n0038_iso_wd_8601-1_2016-02-16.pdf) (page 19), I see that we should provide a leading "P" to the duration specification like "P2Y" for 2 years. In this case, pandas fails to parse that interval reporting:

```
ValueError: Invalid ISO 8601 Duration format - P2Y
```

In this case, I believe pandas is wrong and the ISO spec is right, but that doesn't help us fix the problem. I'll have to think about this a bit more and look for alternative solutions.

huddlej commented:
For what it's worth, [the isodate package](https://github.com/gweis/isodate) has the correct behavior with the above example:


```python
import isodate

>>> isodate.parse_duration("P2Y")
isodate.duration.Duration(0, 0, 0, years=2, months=0)
```

For some reason, I have this package installed already, but I don't know yet what other Nextstrain dependency installs it.

One annoying aspect of the ISO 8601 format is the requirement to provide a leading "P" for the date period values. For Augur, I don't see us needing to support the time period values of hours, minutes, and seconds, so maybe we can automatically prefix a user's given interval string with a "P" if they don't provide that as part of the string.

benjaminotter commented:
ISO8601 duration string parsing with  isodate does the job, but leaves a Deprecation Warning:
```
DeprecationWarning: an integer is required (got type decimal.Decimal).  
Implicit conversion to integers using __int__ is deprecated, and may be 
removed in a future version of Python.
    newdt = other.replace(year=newyear, month=newmonth, day=newday)
```

Also, the isodate package was by default not installed on my local Nextstrain build.

I added a check to make sure the `max-date-offset` and `min_date_offset` strings start with a "P" to prevent failing when the user does not provide the prefix letter.

huddlej commented:
I wanted to revisit this issue, since we still need to add this date offsets feature for ncov and seasonal-flu workflows. I looked into several alternatives for representing date durations with and without support for ISO-8601 parsing including [Arrow](https://github.com/arrow-py/arrow), Python’s built-in datetime timedelta, [isodate](https://github.com/gweis/isodate), [isoduration](https://github.com/bolsote/isoduration), and pandas. Below are some example outputs from each of these for durations from common use cases of 2 years, 6 months, and 4 weeks.

```python
import arrow
import datetime
import isodate
import isoduration
import pandas as pd

# Examples of how to get time delta objects for:
# - 2 years
# - 6 months
# - 4 weeks

# datetime built-in only supports weeks or smaller units
# of time. Years and months have to be converted to the
# closest equivalent in weeks.
#
# 2 years
datetime.timedelta(weeks=52*2)
# datetime.timedelta(days=728)
# 6 months
datetime.timedelta(weeks=int(52*0.5))
# datetime.timedelta(days=182)
# 4 weeks
datetime.timedelta(weeks=4)
# datetime.timedelta(days=28)

# pandas
# 2 years
pd.Timedelta("-2Y")
# FutureWarning: Units 'M', 'Y' and 'y' do not represent unambiguous timedelta values and will be removed in a future version
# Timedelta('-731 days +12:21:36')
#
# pandas generally does not want to support these relative
# offset types. For more details see:
# https://github.com/pandas-dev/pandas/issues/16344
#
# 6 months
pd.Timedelta("-6M")
# FutureWarning: Units 'M', 'Y' and 'y' do not represent unambiguous timedelta values and will be removed in a future version
# Timedelta('-1 days +23:54:00')
# 4 weeks
pd.Timedelta("-4W")
# Timedelta('-28 days +00:00:00')

# arrow does not have a representation of a duration type,
# so we need to create a date object and then "shift" it.
d = arrow.now()
d
# <Arrow [2021-12-29T13:32:46.801604-08:00]>
#
# 2 years
d.shift(years=-2)
# <Arrow [2019-12-29T13:32:46.801604-08:00]>
# 6 months
d.shift(months=-6)
# <Arrow [2021-06-29T13:32:46.801604-07:00]>
# 4 weeks
d.shift(weeks=-4)
# <Arrow [2021-12-01T13:32:46.801604-08:00]>

# isodate was originally developed by one person and has
# had 18 total contributors. It has also been forked into
# a GitHub organization called "isodate" without a clear
# reason. It provides a `parse_duration` function that 
# returns a `Duration` instance that can be added to other
# standard datetime objects. Its parser tries to adhere to
# the ISO-8601 standard completely, such that "2Y" is not
# valid but "P2Y" is. This library only depends on Python's
# datetime library.
#
# 2 years
isodate.parse_duration("-P2Y")
# isodate.duration.Duration(0, 0, 0, years=-2, months=0)
# 6 months
isodate.parse_duration("-P6M")
# isodate.duration.Duration(0, 0, 0, years=0, months=-6)
# 4 weeks
isodate.parse_duration("-P4W")
# datetime.timedelta(days=-28)

# isoduration reimplements the logic from isodate using
# Python's datetime library and the third-party arrow library.
# isoduration claims to address limitations of isodate. Its 
# `parse_duration` function returns a `Duration` instance.
#
# 2 years
isoduration.parse_duration("-P2Y")
# Duration(DateDuration(years=Decimal('-2'), months=Decimal('0'), days=Decimal('0'), weeks=Decimal('0')), TimeDuration(hours=Decimal('0'), minutes=Decimal('0'), seconds=Decimal('0')))
# 6 months
isoduration.parse_duration("-P6M")
# Duration(DateDuration(years=Decimal('0'), months=Decimal('-6'), days=Decimal('0'), weeks=Decimal('0')), TimeDuration(hours=Decimal('0'), minutes=Decimal('0'), seconds=Decimal('0')))
# 4 weeks
isoduration.parse_duration("-P4W")
# Duration(DateDuration(years=Decimal('0'), months=Decimal('0'), days=Decimal('0'), weeks=Decimal('-4')), TimeDuration(hours=Decimal('0'), minutes=Decimal('0'), seconds=Decimal('0')))
```

If we implemented our own minimal duration parsing for Augur to support year, month, week, and day offsets, we would still likely need to implement much of [the algorithm to add durations to datetimes](https://www.w3.org/TR/xmlschema-2/#adding-durations-to-dateTimes) that [isodate already implements](https://github.com/gweis/isodate/blob/d0eb6b98c11a0ed15549b12c39dcc9136727b9e7/src/isodate/duration.py#L60-L61). There is also always the option of straight-up copying the isodate logic into Augur and using the bits we need.

Ultimately, we want to support an interface like this:

```
# Select strains collected no earlier than 2 months ago and no later than 1 day ago (for SARS-CoV-2 analysis).
augur filter --min-date-offset "2M" --max-date-offset "1D"

# Select strains collected no earlier than 2 years ago (for influenza A/H3N2 analysis).
augur filter --min-date-offset "2Y"
```

We don’t especially care about the full ISO-8601 duration implementation (e.g., a duration like “-P2Y3M4D” is unlikely to be useful). Given this simpler interface we need, we could implement durations with `datetime.timedelta` using built-in support for weeks and days and defining our own approximations of years and months. This would get messy, though, and produce odd results for larger durations like 12 years or 11 months where the mismatch between number of weeks in a year or month could add up (52 weeks = 364 days in Python’s timedelta logic!). For example:

```python
now = datetime.datetime.now()
# datetime.datetime(2021, 12, 29, 13, 51, 52, 541769)

# Subtract 1 year in weeks.
now - datetime.timedelta(weeks=52)
# datetime.datetime(2020, 12, 30, 13, 51, 52, 541769)
# Result is off by 1 day.

# Subtract 2 years in weeks.
now - datetime.timedelta(weeks=104)
# datetime.datetime(2020, 1, 1, 13, 51, 52, 541769)
# Result is off by 3 days.

# Subtract 12 years in weeks.
now - datetime.timedelta(weeks=12*52)
# datetime.datetime(2010, 1, 13, 13, 51, 52, 541769)
# Result is off by 2 weeks!
```

Maybe we don’t care that the results are so far off, though? In the course of 12 years, what is 2 weeks in error for the resulting analysis?

After playing with these different implementations a bit, I’m leaning toward adding isodate as a dependency and rebasing the code from this into the new filter interface.

victorlin commented:
@huddlej if using `isodate`/`datetime`, [this won't work with negative dates](https://discuss.python.org/t/bc-date-support/582), for example in [this discussion post](https://discussion.nextstrain.org/t/including-ancient-genomes-with-dates-before-0/516). Would that be appropriate if documented as a feature gap, or should we rethink implementation for this edge case?

huddlej commented:
@victorlin Good point. I'm still not convinced we need to treat negative dates as a first-class use case (as I've mentioned in our Slack discussion). This feature in particular supports the kind of filtering for recent samples we need in the seasonal flu and ncov workflows and that might be strange to implement in an ancient pathogen analysis.

That said, I don't think the proposed interface prevents us from adding support for negative dates in the future, if it becomes necessary. For example, you could run the following to filter to samples collected 10,000 years ago:

```
augur filter --min-date-offset "10000Y"
```

Then it would be on us to update the augur filter internals to support this logic.

If folks are generally on board with the proposed UI, we can get a lot of value now from the isodate implementation.

victorlin commented:
I can look into updating this, since I'll need to incorporate it into #854.

victorlin commented:
There were merge conflicts with 4877de06b189d1250fcf6ce7b88f1bec2705fb86 and 87ca73c49fa35a8e9d3a74f41a0a24ccab5941ea. Force-pushing a rebase on latest `master` which puts the current changes in the right places. Still need to add testing.

trvrb commented:
Really nice to have this implemented @victorlin. I believe you were working with `--min-date-offset` and `--max-date-offset` as initial proposal from @benjaminotter? I would suggest different syntax. If I didn't come in with context that these are meant as dates relative to today I wouldn't know what this refers to. My guess would have actually been that if you supply say `--min-date` of `2021-01-01` that then `--min-date-offset` of `1W` would apply on top of `2021-01-01` resulting in `2021-01-08` (yes, forwards in time would be the expectation).

I don't know if this creates too much ambiguity in terms of parsing, but my preferred user behavior (or at least what seems the most obvious to me) is that there is only `--min-date` that can take `2021-01-01` to have minimal date set to Jan 1, 2021 or could take `1Y` to set minimal date back 1 year from today for `2021-03-29`.

Maximum dates would work similarly. You could have a pair that specifies `--min-date` of `2021-01-01` and `--max-date` of `1W` to give a span from `2021-01-01` to `2022-03-22`.

victorlin commented:
> @trvrb: I believe you were working with --min-date-offset and --max-date-offset as initial proposal from @benjaminotter?

Yes, most of the implementation was already done, I just fixed the merge conflicts and added tests. I also found user experience a bit odd, in that it is a positive duration translating to a negative offset from `date.today()`. But positive offset from `date.today()` (a future time) wouldn't make sense so I figured this is fine, if we want to keep it as offsets from `date.today()` strictly.

So I think there's a question of if we want to support offsets from a certain date, for example:

1. `--min-date 2021-01-01 --min-date-offset 1W` == `--min-date 2021-01-08`
2. `--min-date 2021-01-01 --min-date-offset -1W` == `--min-date 2020-12-25`

If not (keeping offsets strictly from `date.today()`), then it might make more sense to overload `--min-date` to support ISO durations in addition to ISO dates, numeric dates, ambiguous dates, etc.

I can imagine most use cases would offset from `date.today()`. There's no point in hardcoding `--min-date 2021-01-01 --min-date-offset 1W` since that obviously translates to `--min-date 2021-01-08`. If we do want to support offsets from a certain date, it would be up to the user or encompassing workflow to fill in today for `--min-date <today> --min-date-offset 1W`.

huddlej commented:
Thank you for the comments, all! [The original UI I envisioned for this](https://github.com/nextstrain/augur/issues/721) was the overloaded min/max date arguments. When it came time to implement that overloading, we ran into the minor issue that [min/max date arguments are implemented in argparse](https://github.com/nextstrain/augur/blob/945b49af50cc72c4343673491531ae630db40040/augur/filter.py#L1112-L1113) as [a custom numeric_date type](https://github.com/nextstrain/augur/blob/945b49af50cc72c4343673491531ae630db40040/augur/filter.py#L1688-L1703) that would need to be extended or replaced to support the overloading we'd want.

Replacing `numeric_date` felt like asking too much of Benjamin at the time (getting into the weeds of Augur's inconsistent date handling), when the goal of this PR was a quick feature prototype. If the consensus is for the overloaded min/max args (which it seems to be, so far), we just need to handle that `numeric_date` type update.

Regarding this issue:

> If I didn't come in with context that these are meant as dates relative to today I wouldn't know what this refers to.

I don't think any of the proposed syntaxes so far address this potential user confusion. Even the overloaded min/max arguments don't explicitly communicate that the offsets are relative to the current date. The original plan to handle this by clearly describing the offsets in the help text. I also thought the separate offset args might help call out the different behavior from min/max date arguments, but maybe there's a better way...

trvrb commented:
Thanks for context here @huddlej. Very helpful.

I still think that `--max-date 1W` is pretty intuitive in wanting the maximum date to be 1 week from today, but I take your point.

> The original plan to handle this by clearly describing the offsets in the help text. I also thought the separate offset args might help call out the different behavior from min/max date arguments, but maybe there's a better way...

I think I prefer overloading `--min-date` and `--max-date`, but it was really just the word "offset" in `--min-date-offset` that was throwing me above. If the separate commands were `--min-date-from-today` and `--max-date-from-today` I wouldn't have had this confusion.

victorlin commented:
For `--max-date-from-today` I still feel it might take a bit of intuition to realize that it's before today, not after. Ideally it should read something like "max date is 1 week ago". What about `--max-date-ago 1W` to imply temporal direction?

trvrb commented:
Also, as @huddlej notes here https://github.com/nextstrain/augur/issues/721#issuecomment-864186135, we should plan to use the same interface for `augur frequencies`, which also currently takes `--min-date` and `--max-date`. I obviously wouldn't wrap changes to `augur frequencies` into this PR, but writing here as a reminder about this. For the planned `ncov` partitioning into `6m` and `all-time` targets, it would be useful to have this interface also in place for `augur frequencies`.

victorlin commented:
Ok, I'll update PR to overload `--min-date` and `--max-date`.

victorlin commented:
Note: I've reworded the PR and documentation to use the term **_relative date_** instead of **_date offset_** because I think it helps in explaining this feature as an add-on to previously supported formats of **_absolute date_**.

huddlej reviewed (commented):
This looks great, @victorlin, especially the numerous tests! I only had minor comments threaded below.

I tested this out with the ncov workflow using the following build config:

```yaml
inputs:
  - name: botswana-data
    metadata: data/gisaid_auspice_input_hcov-19_2022_03_29_17.tar
    sequences: data/gisaid_auspice_input_hcov-19_2022_03_29_17.tar
  - name: references
    metadata: data/references_metadata.tsv
    sequences: data/references_sequences.fasta

builds:
  botswana:
    subsampling_scheme: recent

filter:
  skip_diagnostics: true
  min_length: 5000

subsampling:
  recent:
    recent_focal:
      min_date: --min-date 1M
    older_focal:
      min_date: --min-date 2M
      max_date: --max-date 1M
```

Everything worked as I expected with the example data from Botswana.

victorlin commented on augur/filter.py:
```diff
@@ -1689,18 +1697,46 @@ def numeric_date(date):
+        else:
+            duration_str = 'P'+duration_str
+        return treetime.utils.numeric_date(datetime.date.today() - isodate.parse_duration(duration_str))
+    except:
```
Good point, added: 5fbe597a

Found `ValueError` and `isodate.ISO8601Error` using these examples:

```py
import datetime
date = 'XXXX-02-03'
datetime.date(*map(int, date.split("-", 2)))
# ValueError: invalid literal for int() with base 10: 'XXXX'

date = '-XXXX-02-03'
datetime.date(*map(int, date.split("-", 2)))
# ValueError: invalid literal for int() with base 10: ''

date = 'invalid'
datetime.date(*map(int, date.split("-", 2)))
# ValueError: invalid literal for int() with base 10: 'invalid'

date = 'P3000Y'
datetime.date(*map(int, date.split("-", 2)))
# ValueError: invalid literal for int() with base 10: 'P3000Y'

import isodate
duration_str = 'P3000Y'
datetime.date.today() - isodate.parse_duration(duration_str)
# ValueError: year -978 is out of range

duration_str = 'invalid'
datetime.date.today() - isodate.parse_duration(duration_str)
# ISO8601Error: Unable to parse duration string 'invalid'

duration_str = '-XXXX-02-03'
datetime.date.today() - isodate.parse_duration(duration_str)
# ISO8601Error: Unable to parse duration string '-XXXX-02-03'
```

victorlin commented on setup.py:
```diff
@@ -53,6 +53,7 @@
     install_requires = [
         "bcbio-gff >=0.6.0, ==0.6.*",
         "biopython >=1.67, !=1.77, !=1.78",
+        "isodate ==0.6.*",
```
Added as task in the PR description.

victorlin commented on tests/test_filter.py:
```diff
@@ -265,3 +267,133 @@ def test_filter_date_formats(self, tmpdir, fasta_fn, argparser):
         output = SeqIO.to_dict(SeqIO.parse(out_fn, "fasta"))
         assert list(output.keys()) == ["SEQ_1", "SEQ_2", "SEQ_3"]
+
+    def test_filter_min_date_relative_1m(self, tmpdir, argparser, patch_datetime_today):
```
Thanks for the pointer! [Done](https://github.com/nextstrain/augur/compare/c726b5fdcdd194f67d9d6464b6e9a0ff83cc7a37..19c3eb717e706052969cc0ccf76a9787d0b5c67a).

joverlee521 commented on tests/test_dates.py:
```diff
@@ -0,0 +1,17 @@
+FIXED_DATE = datetime.date(2020, 3, 25)
+
+@pytest.fixture
+def patch_datetime_today(monkeypatch):
+    class FixedDate(datetime.date):
+        @classmethod
+        def today(cls):
+            return FIXED_DATE
+    monkeypatch.setattr(datetime, 'date', FixedDate)
```
This seems to be doing the same thing as `freeze_time` used in [test_date_disambiguator](https://github.com/nextstrain/augur/blob/master/tests/util_support/test_date_disambiguator.py)?

victorlin replied:
Didn't realize `freezegun` was already used here. Thanks for the pointer!

Updated to use this, and have consistent styling for `pytest.mark.parametrize`: [diff](https://github.com/nextstrain/augur/compare/8758ad25..a4620431)

joverlee521 commented on augur/filter.py:
```diff
@@ -319,9 +319,8 @@ def filter_by_date(metadata, date_column="date", min_date=None, max_date=None):
 
-    # Skip this filter if no valid min/max date is given or the date column does
-    # not exist.
-    if (not min_date and not max_date) or date_column not in metadata.columns:
```
Removing this check does slightly change the function, i.e. it will now filter out strains with invalid dates even when min/max date is not provided. Maybe also update the docstring & doctests to reflect this change?

victorlin replied:
Thinking about it some more, I think the checks for `args.min_date`/`args.max_date` and `date_column` should be done in the same place. Ideally would be [here](https://github.com/nextstrain/augur/blob/945b49af50cc72c4343673491531ae630db40040/augur/filter.py#L609), but `metadata.columns` is not available with the current implementation of `construct_filters()`.

This change was needed in the original implementation using `args.min_date_offset`/`args.max_date_offset` and optional for the current implementation.

I'll drop fbb1303169efa5853d67133b44454f81e5454b04 for now.

joverlee521 commented on augur/filter.py:
```diff
@@ -1735,6 +1735,10 @@ def numeric_date(date):
 
+    # This currently doesn't get exposed since argparse raises a SystemExit on invalid arguments.
+    # TODO: find a way to provide better errors for invalid dates.
+    raise ValueError(f"Unable to determine date from {date}.")
```
Error messages passed to `argparse.ArgumentTypeError` replace the generic type error and gets surfaced to the user. ( I forget where I read this from...maybe an old stackoverflow?)

victorlin replied:
Oh cool! This fixes from a functional perspective.

Console output with `raise ValueError`:

```
augur filter: error: argument --max-date: invalid numeric_date value: '3000Y'
```

With `raise argparse.ArgumentTypeError`:

```
augur filter: error: argument --max-date: Unable to determine date from 3000Y.
```

However, both surface a `SystemExit` without the actual message in the pytest. Maybe a cram test is more suitable here.

joverlee521 replied:
Ah, right the pytest. This should be in the SystemExit context, see https://stackoverflow.com/questions/49323496/check-argparse-argumenttypeerror

victorlin replied:
Ohh the message is all the way down in `ExceptionInfo.value.__context__.message`.

Pushed an update: a2ac8e5a

joverlee521 commented on tests/test_filter.py:
```diff
@@ -265,3 +268,114 @@ def test_filter_date_formats(self, tmpdir, fasta_fn, argparser):
+    @freeze_time("2020-03-25")
+    @pytest.mark.parametrize(
+        "argparse_flag, argparse_value",
+        [
+            ("--max-date", "3000Y"),
+            ("--max-date", "invalid"),
+        ],
```
Minor nitpick, please add tests for invalid `--min-date` so we cover all bases.

joverlee521 reviewed (approved):
Looks good! Thanks for addressing my sporadic comments!

victorlin commented:
Thanks for the review @joverlee521! Force-pushing a small test order fix then merging.

victorlin commented:
I'm deleting an old branch `filter-date-offsets` which strays from `master` by 1 commit: 00ec319534331c928985832f61fce76170f3fab0

It isn't associated with any closed PR and looks like an earlier implementation of what got merged in this PR, so I'm linking it here for posterity.
